src/layers/core/grid-layer/grid-layer-vertex.glsl

```diff
@@ -43,41 +29,42 @@ varying vec4 vColor;
 
 void main(void) {
 
-
   // cube gemoetry vertics are between -1 to 1, scale and transform it to between 0, 1
-  vec2 ptPosition = instancePositions.xy + vec2((positions.x + 1.0 ) * lonOffset / 2., (positions.y + 1.) * latOffset / 2.);
+  vec2 ptPosition = instancePositions.xy + vec2((positions.x + 1.0 ) * lonOffset / 2.0, (positions.y + 1.0) * latOffset / 2.0);
 
   vec2 pos = project_position(ptPosition);
 
-  // if 3d not enabled set elevation to 0
-  float elevation =  mix(0., project_scale(instancePositions.w  * (positions.z + 1.) * ELEVATION_SCALE), extruded);
+  float elevation = 0.0;
+
+  if (extruded > 0.5) {
+    elevation = project_scale(instancePositions.w  * (positions.z + 1.0) * ELEVATION_SCALE);
+  }
 
   // extrude positions
-  vec3 extrudedPosition = vec3(pos.xy, elevation + 1.);
+  vec3 extrudedPosition = vec3(pos.xy, elevation + 1.0);
+  vec4 position_worldspace = vec4(extrudedPosition, 1.0);
+  gl_Position = project_to_clipspace(position_worldspace);
+
+  if (renderPickingBuffer < 0.5) {
+    float selected = isPicked(instancePickingColors, selectedPickingColor);
 
-  float selected = isPicked(instancePickingColors, selectedPickingColor);
+    float lightWeight = 1.0;
 
-  vec3 lightWeighting = getLightWeight(
-    viewMatrix,
-    worldMatrix,
-    worldInverseTransposeMatrix,
-    positions,
-    normals,
-    selected,
-    ambientColor,
-    pointLocation,
-    pointColor,
-    pointSpecularColor,
-    pointLightAmbientCoefficient,
-    shininess
-  );
+    if (extruded > 0.5) {
+      lightWeight = getLightWeight(
+        position_worldspace,
+        normals
+      );
+    }
 
-  vec3 lightWeightedColor = mix(vec3(1.), lightWeighting, extruded) * (instanceColors.rgb / 255.);
-  vec4 color = vec4(lightWeightedColor, instanceColors.a * opacity / 255.);
+    vec3 lightWeightedColor = lightWeight * (instanceColors.rgb / 255.);
+    vec4 color = vec4(lightWeightedColor, instanceColors.a * opacity / 255.);
+    vColor = color;
 
-  vec4 pickingColor = vec4(instancePickingColors / 255., 1.);
+  } else {
 
-  vColor = mix(color, pickingColor, renderPickingBuffer);
```

Comment:
I'm benchmarking mix() vs if() right now. I hope to get a solid conclusion on this today.

However, at least we should remove this mix because if we are rendering picking buffer, we should skip the expansive lighting calculations for sure.